### ISSUE TITLE
Matrix doesn't rref for floating point numbers

### ISSUE THREAD

**Issue by typedeph** [sympy/sympy#9060]
The following Matrix doesn't row reduce correctly
```
A = sympy.Matrix([
    [0.9, -0.1, -0.2, 0],
    [-0.8, 0.9, -0.4, 0],
    [-0.1, -0.8, 0.6, 0]
])
A.rref()
# Output
(Matrix([
 [1.0,   0,   0, 0],
 [  0, 1.0,   0, 0],
 [  0,   0, 1.0, 0]]), [0, 1, 2])
```
While in Matlab the result is
```
A = [0.9 -0.1 -0.2 0; -0.8 0.9 -0.4 0; -0.1 -0.8 0.6 0]
rref(A)
% Output
ans =

    1.0000         0   -0.3014         0
         0    1.0000   -0.7123         0
         0             0             0         0
```

**Comment by cbm755** [comment]
Yes, the SymPy implementation is not robust in floating point.  Fixing this would be a nice exercise for a numerical analysis student.

1.  it is numerically unstable (b/c of the pivoting strategy).  But that is not the problem here.
2.  the various row operations will leave machine epsilon in `A[2,2]` and there is no tolerance in the determining of rank.

Here is a simpler example:
````
A = Matrix([[1.0, 0.0],[0.0, 1e-16]])
A.rref()

Out[61]: 
(Matrix([
 [1.0,   0],
 [  0, 1.0]]), [0, 1])
````

Note as you probably know, your example works if done symbolically:
````
In [52]: B = Matrix([[9, -1, -2, 0], [-8, 9, -4, 0], [-1, -8, 6, 0]])

In [53]: B.rref()
Out[53]: 
(Matrix([
 [1, 0, -22/73, 0],
 [0, 1, -52/73, 0],
 [0, 0,      0, 0]]), [0, 1])
````

**Comment by smichr** [comment]
If you have an idea of how small a number should be before it is no longer significant you can use the iszerofunc keyword for rref:

```
>>> A.rref(iszerofunc=lambda x:comp(x,0,1e-9))
(Matrix([
[1.0,   0,    -0.301369863013699, 0],
[  0, 1.0,    -0.712328767123288, 0],
[  0,   0, -1.11022302462516e-16, 0]]), [0, 1])
>>> _[0].applyfunc(lambda x: x.n(3,chop=1e-9))
Matrix([
[1.0,   0, -0.301, 0],
[  0, 1.0, -0.712, 0],
[  0,   0,      0, 0]])
```

**Comment by paramsingh** [comment]
How should this be fixed?

**Comment by cbm755** [comment]
Use NumPy...  j/k, mostly :)

Serious answer: it would be great to be able to do serious high precision floating point so it would be good to fix this.

1. implement partial pivoting (max of column, not just first nonzero like I think it does now).  This gives numerical stability [for almost all matrices](https://people.maths.ox.ac.uk/trefethen/nov12.pdf).

2. perhaps @smichr's suggestion can be done automatically based on the precision of the `Float`s inside the `Matrix`?
    (a)  Look at the `rref` code in GNU Octave for example.  Or check NumPy.
    (b)  note whatever they use for defaults is based on machine epsilon 1e-16.

**Comment by cbm755** [comment]
Item 1 in particular should be a very easy fix.

**Comment by Jawadh-Salih** [comment]
@paramsingh, are you done with this patch?

**Comment by paramsingh** [comment]
I'm gonna start work on it today most probably, @Jawadh-Salih

**Comment by ahhda** [comment]
This is my first bug fix. The code may not look nice.
I have Implemented partial pivoting. Please check and suggest necessary changes.

**Comment by debugger22** [comment]
@ahhda Where is your fix?

**Comment by ahhda** [comment]
@debugger22 It failed some tests, I will create a pull request once I solve those.

**Comment by cbm755** [comment]
Its also ok to leave PRs open while working on things.  You can mark them [WIP].

**Comment by ahhda** [comment]
I have created a pull request. [#9109] (https://github.com/sympy/sympy/pull/9109).
On running tests I get the error "Can't evaluate eigenvector for eigenvalue (some integer)".
How should I go about this ?

**Comment by smichr** [comment]
> Item 1 in particular should be a very easy fix.

I warn that this is a hollow victory since even with partial pivoting, a singular matrix will "succeed" with the new method because a virtual zero will not be detected as zero. To make this robust one needs to know whether a number is actually zero or not. If it isn't zero then no matter how small it is SymPy should work, right? SymPy's issue isn't numerical stability, it's zero detection. My concern is that s *symbolic* routine is being modified to do numerics better. This is a problem faced in solving equations, too. The approach taken there is to recast in Rationals and then recast the solution in Floats. I am not sure what the best approach is for matrices. I do not believe we should be implementing partial pivoting, however.

**Comment by cbm755** [comment]
@smichr I guess I don't really know the SymPy philosophy for Floats yet.  But currently the algorithm runs effectively in fixed precision (I think its the best---or maybe worst---precision of the Floats involved) so improving the numerical stability should decrease the growth in error.

Put another way: with an unstable algorithm, a person cannot know what to put for iszerofunc (see your coment https://github.com/sympy/sympy/issues/9060#issuecomment-76320507).  With a stable algorithm, she has some hope...

I'm not debating your broader point that there might be a more symbolic fix.  (My use-case for Floats is to pretend I have a more accurate but still fixed-precision IEEE floating point system).

**Comment by Shekharrajak** [comment]
Is this issue fixed ?

**Comment by typedeph** [comment]
I just retried my example and yes, the rref output is still the same as when I first posted the issue. As smichr stated though, if you use iszerofunc or fractions you can get the result you desire.

**Comment by ashutoshsaboo** [comment]
Is this solved? I would like to work towards solving this? Can I work on this issue, in case if it's not solved till now?

**Comment by cbm755** [comment]
It is not solved.  Perhaps "Easy-to-Fix" was not appropriate.

If you want to help, see the PR #9060 and #9121.  @ahhda I'm sorry I haven't been keeping up with this!

**Comment by cbm755** [comment]
@siefkenj sorry I haven't been following your work but this has been fixed by #11554 right?

**Comment by siefkenj** [comment]
@cbm755 No, this has not been fixed yet.  To fix this issue automatic detection of the condition number for matrices with Floats is needed.  I would suggest that this issue be fixed by a new `FloatMatrix` subclass and that the matrix constructor automatically would cast to a `FlaotMatrix` if every entry was a float or zero.

**Comment by smichr** [comment]
> SymPy's issue isn't numerical stability, it's zero detection.

Alas, there are numerical issues as well. Accumulation of round-off error can still plague calculations as demonstrated [here](https://github.com/sympy/sympy/issues/17136#issuecomment-508769239).
